clients/src/main/java/org/apache/kafka/clients/consumer/internals/OffsetsFinder.java

```diff
@@ -0,0 +1,805 @@
+    private final ConsumerNetworkClient client;
+    private final Time time;
+    private final long retryBackoffMs;
+    private final long requestTimeoutMs;
+    private final ConsumerMetadata metadata;
+    private final SubscriptionState subscriptions;
+    private final IsolationLevel isolationLevel;
+    private final AtomicReference<RuntimeException> cachedListOffsetsException = new AtomicReference<>();
+    private final AtomicReference<RuntimeException> cachedOffsetForLeaderException = new AtomicReference<>();
+    private final OffsetsForLeaderEpochClient offsetsForLeaderEpochClient;
+    private final ApiVersions apiVersions;
+    private final AtomicInteger metadataUpdateVersion = new AtomicInteger(-1);
+
+    public OffsetsFinder(LogContext logContext,
+                           ConsumerNetworkClient client,
+                           ConsumerMetadata metadata,
+                           SubscriptionState subscriptions,
+                           Time time,
+                           long retryBackoffMs,
+                           long requestTimeoutMs,
+                           IsolationLevel isolationLevel,
+                           ApiVersions apiVersions) {
+        this.log = logContext.logger(Fetcher.class);
+        this.time = time;
+        this.client = client;
+        this.metadata = metadata;
+        this.subscriptions = subscriptions;
+        this.retryBackoffMs = retryBackoffMs;
+        this.requestTimeoutMs = requestTimeoutMs;
+        this.isolationLevel = isolationLevel;
+        this.apiVersions = apiVersions;
+        this.offsetsForLeaderEpochClient = new OffsetsForLeaderEpochClient(client, logContext);
+    }
+
+    /**
+     * Represents data about an offset returned by a broker.
+     */
+    static class ListOffsetData {
+        final long offset;
+        final Long timestamp; //  null if the broker does not support returning timestamps
+        final Optional<Integer> leaderEpoch; // empty if the leader epoch is not known
+
+        ListOffsetData(long offset, Long timestamp, Optional<Integer> leaderEpoch) {
+            this.offset = offset;
+            this.timestamp = timestamp;
+            this.leaderEpoch = leaderEpoch;
+        }
+    }
+
+    /**
+     * Get topic metadata for all topics in the cluster
+     * @param timer Timer bounding how long this method can block
+     * @return The map of topics with their partition information
+     */
+    public Map<String, List<PartitionInfo>> getAllTopicMetadata(Timer timer) {
+        return getTopicMetadata(MetadataRequest.Builder.allTopics(), timer);
+    }
+
+    /**
+     * Get metadata for all topics present in Kafka cluster
+     *
+     * @param request The MetadataRequest to send
+     * @param timer Timer bounding how long this method can block
+     * @return The map of topics with their partition information
+     */
+    public Map<String, List<PartitionInfo>> getTopicMetadata(MetadataRequest.Builder request, Timer timer) {
```

Comment:
This doesn't seem related to finding offsets. Should it be somewhere else?

Reply:
Agreed. Is adding it to `KafkaConsumer` directly OK?

Reply:
We can generalize the name perhaps? Instead of `OffsetFinder`, perhaps we can call it `MetadataManager` or something generic like that.

Reply:
Name has been changed on the latest commit.